### ISSUE TITLE
Support scalar tensors

### ISSUE THREAD

**Issue by yoheiueda** [torch-spyre/torch-spyre#1334]
### What feature would you like to be added?

A scalar tensor in PyTorch is a tensor that contains exactly one value and has no dimensions.

`torch.tensor(5)` is a **scalar tensor** (a 0-D tensor with shape `[]`), while `torch.tensor([5])` is a **1-D tensor with one element** (shape `[1]`).

In other words:

- Python scalar: `5`
- Scalar tensor: `torch.tensor(5)`
- Size-1 1-D tensor: `torch.tensor([5])`

Simple comparison:

| Kind | Shape | Meaning |
|---|---|---|
| Python scalar | N/A | Plain Python number |
| Scalar tensor | `[]` | Single value stored as a tensor |
| 1-D size-1 tensor | `[1]` | Vector with one element |

So even though `torch.tensor(5)` and `torch.tensor([5])` both contain one value, they are **not the same shape** and can behave differently in shape-sensitive operations.

### Why is this feature needed?

We need to support scalar tensors for enablement of various operators:

- https://github.com/torch-spyre/torch-spyre/issues/1172
- https://github.com/torch-spyre/torch-spyre/issues/1327
- https://github.com/torch-spyre/torch-spyre/issues/1187
- https://github.com/torch-spyre/torch-spyre/issues/1188

### Anything else we need to know?

Currently, constant arguments are represented as size-1 1-D tensors. Once scalar tensor support is available, we can switch to representing them as scalar tensors instead.

- https://github.com/torch-spyre/torch-spyre/issues/238
- https://github.com/torch-spyre/torch-spyre/pull/915
- https://github.com/torch-spyre/torch-spyre/issues/1051

**Comment by yoheiueda** [comment]
In my opinion, a scalar tensor should be represented as 0-D host tensor metadata together with a 1-D size-1 device tensor layout, as shown below:

 | Host size  | Host size | Device size |
| ------------- | ------------- | ------------- |
| Scalar tensor (option 1) | []  | [1, 64] or [1, -1] |
| Scalar tensor (option 2)| []  | [64] or [-1] |
| 1-D size-1 dense tensor | [1]  | [1, 64] |
| 1-D size-1 sparse tensor | [1]  | [1, -1] |


However, the current runtime implementation converts a 0-D tensor to a 1-D tensor.

```
>>> x=torch.tensor(123.0, dtype=torch.float16)
>>> x.size()
torch.Size([])
>>> y=x.to("spyre")
>>> y.size()
torch.Size([1])
>>> y.device_tensor_layout().device_size
[1, 64]
```

This conversion appears to happen in the following lines of `spyre_empty_strided`:
https://github.com/torch-spyre/torch-spyre/blob/182e4356a59bf5881724575b4bade84b020c7740/torch_spyre/csrc/spyre_mem.cpp#L531-L541

To fully support scalar tensors, I think we need updates on both the frontend runtime side and the compiler side.

On the frontend compiler side, we need to update logic that references host sizes together with device sizes so that scalar tensor shapes are handled correctly. This likely affects code such as `views.py` and `stickify.py`.

I’m not sure whether option 1 or option 2 is the better choice here. Option 2 preserves the invariant `len(host_size) == len(device_size) - 1`, but a lot of the existing code assumes `len(device_size) >= 2`.  It would also require updates to the codegen part like `superdsc.py`.

On the runtime side, it probably makes sense to wait for the ongoing runtime refactoring to settle, and then revisit this once the new implementation is ready.

For testing ops before runtime support is available, one possible workaround is to use `x.unsqueeze(dim=0)` to convert scalar output tensors into 1-D tensors.

**Comment by moriohara** [comment]
Assigned to @joyalbin

**Comment by imaihal** [comment]
I also discussed before bout this in #1172 with @thoangtrvn. He suggested to use a flag. 

 https://github.com/torch-spyre/torch-spyre/issues/1172#issuecomment-4135115280
> My understanding of PR https://github.com/torch-spyre/torch-spyre/pull/1171 is that it sets the returned tensor’s metadata to scalar on transfer back, but the actual transferred tensor is still treated/produced as [1] internally—so it doesn’t fully resolve the ambiguity at the source.
> Proposed fix direction: the DT/compiler side should avoid erasing scalar-ness when lowering to dimensional form, e.g. by either:
> 
>     - preserving an explicit “was_scalar” flag / scalar intent through the Spyre representation, or
>     - introducing a dedicated scalar representation path so we don’t have to canonicalize [] -> [1] in the first place.
> 
> This would let .to("cpu") faithfully reconstruct shape=[] only when the original CPU tensor was truly scalar, while keeping genuine [1] tensors as [1].

 @thoangtrvn any comments on this?

**Comment by yoheiueda** [comment]
I think the second suggestion from @thoangtrvn is essentially the same as what I described above.

In the frontend compiler and runtime, we already preserve both the host size and the device size, and both can be referenced when needed. Because of that, I don’t think we need an additional flag to distinguish scalar tensors from 1-D size-1 tensors.

**Comment by joyalbin** [comment]
While analyzing different issues with scalar tensor, observed following inconsistencies with scalar tensor representation:

1. Transferring a ‘cpu’ scalar tensor (0-D) to ‘spyre’ is converting the tensor to 1D. This change is causing issues https://github.com/torch-spyre/torch-spyre/issues/1172 and https://github.com/torch-spyre/torch-spyre/issues/1327 as tensor transferring back to cpu is a 1D tensor.
2. Tensor generation operations like torch.empty can produce 0D tensor on spyre device. But for any kernel launch at runtime this 0D tensor is converted to 1D . https://github.com/torch-spyre/torch-spyre/blob/main/torch_spyre/csrc/module.cpp#L188 . This change is causing issue https://github.com/torch-spyre/torch-spyre/issues/1188

Looking for a clarity from https://github.com/torch-spyre/torch-spyre/pull/1665#discussion_r3124372534 that the data transfer and the kernel launch have any constrain that tensor should be atleast 1D (scalar tensor of 0D must convert to 1D)

cc: @thoangtrvn

**Comment by yoheiueda** [comment]
This PR should address the runtime‑side issue.

- #1766

**Comment by joyalbin** [comment]
The issue https://github.com/torch-spyre/torch-spyre/issues/1187 will be address by PR https://github.com/torch-spyre/torch-spyre/pull/1922

**Comment by joyalbin** [comment]
Summary on the issues attached this this EPIC:
1. Issues https://github.com/torch-spyre/torch-spyre/issues/1172 and https://github.com/torch-spyre/torch-spyre/issues/1327  got resolved with PR https://github.com/torch-spyre/torch-spyre/pull/1766
2. Issue https://github.com/torch-spyre/torch-spyre/issues/1188 got resolved with PR https://github.com/torch-spyre/torch-spyre/pull/1823
3. The issue https://github.com/torch-spyre/torch-spyre/issues/1187 fixes with PR  https://github.com/torch-spyre/torch-spyre/pull/1922

**Comment by yoheiueda** [comment]
@joyalbin Thanks for summarizing the status of scalar tensor support! Given that all related issues have been resolved, let’s go ahead and close this issue.

**Comment by yoheiueda** [comment]
Oh, sorry. #1922 is not yet merged.

**Comment by yoheiueda** [comment]
PR #1922 has been merged. Closing this issue.

**Comment by moriohara** [comment]
Closing as the following two test cases passed:
- tests/models/test_model_ops.py::TestSpyreModelOpsPRIVATEUSE1::test_model_ops_db_torch_add__1__Qwen2_5-7B-Instruct_yaml__spyre_float16
- tests/models/test_model_ops.py::TestSpyreModelOpsPRIVATEUSE1::test_model_ops_db_torch_add__1__granite-4_1-8b_yaml__spyre_float16
